### PR TITLE
BarChart: Use new uplot layout for rendering bars

### PR DESCRIPTION
**Pull request by dprokop**
Implementing upcoming improvements to uPlot bars rendering:)

TODO
- [x] Bring back bars fill
- [x] Bring back value label auto sizing - autosizing can be tested in this dashboard: http://localhost:3000/d/WFlOM-jM1/barchart-panel-tests-value-sizing?orgId=1
- [ ] update to latest uPlot when it's released

### CODE REVIEW

File: public/app/plugins/panel/barchart/utils.ts [hunk 1]
```diff
@@ -80,7 +80,7 @@ export const preparePlotConfigBuilder: UPlotConfigPrepFn<BarChartOptions> = ({
 
   builder.addHook('init', config.init);
   builder.addHook('drawClear', config.drawClear);
-  builder.addHook('draw', config.draw);
+  builder.addHook('drawPoints', config.drawPoints);
```
Comment (dprokop) [inline public/app/plugins/panel/barchart/utils.ts:83]: @leeoniya TS is complaining about drawPoints missing from hook defs - has it been removd or we ts-ignored when we used it some time ago?

**Comment by dprokop** [comment]
So @leeoniya i took it to a "renderable" state - still loads of stuff missing, but the major thing that I cannot get to work is the bars fill. I assume(maybe wrong assumption) that the path builder uses uses stroke or fill color, yet the bars are rendered empty - any ideas? Probably some stupid mistake of mine

File: public/app/plugins/panel/barchart/utils.ts (hunk 1 above)
Comment (leeoniya) [reply public/app/plugins/panel/barchart/utils.ts:83]: there's no `drawPoints` hook. drawPoints is just a function to use for `series.points.show`

**Comment by dprokop** [comment]
> So @leeoniya i took it to a "renderable" state - still loads of stuff missing, but the major thing that I cannot get to work is the bars fill. I assume(maybe wrong assumption) that the path builder uses uses stroke or fill color, yet the bars are rendered empty - any ideas? Probably some stupid mistake of mine

@leeoniya  did a debug agains [your demo](https://github.com/leeoniya/uPlot/blob/0f3cbd4ae1442214946d47212a2be02f5e0355d2/demos/multi-bars.html), and it seems like it doesn't like rgba strings as fill 😬

**Comment by dprokop** [comment]
Ok, made it work, or close to work. Now struggling with the sizings being incorrect:

File: public/app/plugins/panel/barchart/bars.ts [hunk 2]
```diff
@@ -108,6 +112,16 @@ export function getConfig(opts: BarsOptions, theme: GrafanaTheme2) {
...
     each: (u, seriesIdx, dataIdx, lft, top, wid, hgt) => {
+      let val = u.data[seriesIdx][dataIdx];
+
+      if (isXHorizontal) {
+        vSpace = Math.min(vSpace, val! < 0 ? u.bbox.top + u.bbox.height - (top + hgt) : top - u.bbox.top);
+        hSpace = wid;
+      } else {
+        vSpace = hgt;
+        hSpace = Math.min(hSpace, val! < 0 ? lft - u.bbox.left : u.bbox.left + u.bbox.width - (lft + wid));
+      }
```
Comment (dprokop) [inline public/app/plugins/panel/barchart/bars.ts:123]: ah very cool!

**Comment by leeoniya** [comment]
i've pushed a pretty sizable refactoring (pardon the pun) that simplifies and optimizes the label measuring machinery.

another simplification is to simply rely on the `draw` hook for measuring and rendering labels. since we need to do this cross-series, this is the logical single place to do it instead of trying shoe-horn label rendering into the point renderer which runs per-series, without cross-series context. hopefully this makes the code more straightforward and understandable.

i did not try to get things pixel-perfect. also, i made a few adjustments that i think look better and space the labels a bit further from the bars. in general everything seems to work well, including label auto-hiding in both orientations as well as properly choosing font sizes efficiently. we may need to hard-code a spacing increase for negative-valued labels on vertical bars, since top-aligning the text does looks to be closer than baseline aligning.

before/after:

https://user-images.githubusercontent.com/43234/126044208-6341fa73-9cef-472d-a69d-4f8e0a21c8e8.mp4

**Comment by leeoniya** [comment]
also discovered that all barcharts fully re-initialize on every data update (before and after this PR)

need to fix that before v8.1 in a follow-up

**Comment by ryantxu** [comment]
This is looking good!  When I mess with it, it seems that the font size depends on both width & height -- does it make sense to have it depend on just one?  and add enough padding that we can always draw the label for the tallest/widest item?  The size shifting seems awkward when we have plenty of space to draw things nicely except for one element

Peek 2021-07-19 22-25

**Review by leeoniya — APPROVED** [review]

**Pull request merged**